clang/lib/Sema/SemaAPINotes.cpp

```diff
@@ -783,51 +783,76 @@ static void ProcessVersionedAPINotes(
   }
 }
 
+static std::optional<api_notes::Context>
+UnwindNamespaceContext(DeclContext *DC, api_notes::APINotesManager &APINotes) {
+  if (auto NamespaceContext = dyn_cast<NamespaceDecl>(DC)) {
+    for (auto Reader : APINotes.findAPINotes(NamespaceContext->getLocation())) {
+      // Retrieve the context ID for the parent namespace of the decl.
+      std::stack<NamespaceDecl *> NamespaceStack;
+      {
+        for (auto CurrentNamespace = NamespaceContext; CurrentNamespace;
+             CurrentNamespace =
+                 dyn_cast<NamespaceDecl>(CurrentNamespace->getParent())) {
+          if (!CurrentNamespace->isInlineNamespace())
+            NamespaceStack.push(CurrentNamespace);
+        }
+      }
+      std::optional<api_notes::ContextID> NamespaceID;
+      while (!NamespaceStack.empty()) {
+        auto CurrentNamespace = NamespaceStack.top();
+        NamespaceStack.pop();
+        NamespaceID =
+            Reader->lookupNamespaceID(CurrentNamespace->getName(), NamespaceID);
+        if (!NamespaceID)
+          return std::nullopt;
+      }
+      if (NamespaceID)
+        return api_notes::Context(*NamespaceID,
+                                  api_notes::ContextKind::Namespace);
+    }
+  }
+  return std::nullopt;
+}
+
+static std::optional<api_notes::Context>
+UnwindTagContext(TagDecl *DC, api_notes::APINotesManager &APINotes) {
+  for (auto Reader : APINotes.findAPINotes(DC->getLocation())) {
+    // Retrieve the context ID for the parent tag of the decl.
+    std::stack<TagDecl *> TagStack;
+    {
+      for (auto CurrentTag = DC; CurrentTag;
+           CurrentTag = dyn_cast<TagDecl>(CurrentTag->getParent()))
+        TagStack.push(CurrentTag);
+    }
+    assert(!TagStack.empty());
```

Comment:
Can we make an assertion earlier that the tag is embedded in at least one namespace?

Reply:
Not sure I follow: a tag might be a top-level entity, if its context is the translation unit. It isn't necessarily nested in a C++ namespace.

Reply:
I'm just wondering if we can hoist this assertion. It is a bit confusing to assert that the tag stack is not empty versus checking where it is declared.

Reply:
Hmm, the reason I added the assertion here is because we're calling `TagStack.top()->...` on the following line, and I wanted to make it clear to the reader that this is access is never going to be out of bounds.

Reply:
Right, but that also implies that the loop above also executed. That means the loop invariant also must hold. The tag stack can never be empty because it is initialized with the tag as the decl context. But you can also work backwards and assert that DC is not null and hoist the assertion. You can also remove the assertion by taking a reference to the DC.

Reply:
Yeah, that makes sense. I added another assertion that ensures `DC` is not null. I would prefer to keep this assertion in place, since I think C++ container access without bounds check tends to make people nervous in non-hardened environments.

> You can also remove the assertion by taking a reference to the DC.

I thought the implicit convention of the codebase is that `clang::ABCDecl` are passed by pointer, not by reference. I could be wrong. But I think it's better to be consistent with the surrounding code.